src/qumin/representations/frequencies.py

```diff
@@ -0,0 +1,398 @@
+                                 usecols=['form', 'value', 'freq_id'])
+            freq_col = freq.columns
+            if "form" not in freq_col:
+                raise ValueError("The form column is required in the frequency table if real=True and no frequencies in the forms table.")
+
+            if "source" not in freq_col:
+                freq['source'] = 'frequencies_table'
+                cls.source['forms'] = 'frequencies_table'
+            elif cls.source['forms'] is None:
+                cls.source['forms'] = list(cls.freq['source'].unique())[0]
+                log.info(f"No default source provided for frequencies. Using {cls.source['forms']}")
+
+            # We use the form_id column to match both dataframes
+            paradigms.set_index('form_id', inplace=True)
+            freq.set_index('form', inplace=True)
+
+            missing_idx = ~paradigms.index.isin(freq.index)
+            if missing_idx.any():
+                log.warning(f"""The frequencies table does not contain a row for every form_id row. Missing:
+                            {paradigms.loc[missing_idx].head()}""")
+
+            paradigms.loc[cls.freq.index, 'value'] = freq.value
+            cls.forms = paradigms.copy()
+
+        else:
+            if real:
+                log.warning("""No form frequencies were available in the Paralex dataset.""")
+            log.info('forms: Building empty frequencies.')
+            cls.forms = paradigms
+            cls.forms['source'] = 'empty'
+            cls.forms['value'] = pd.NA
+            cls.source['forms'] = 'empty'
+
+        # Check for duplicate overabundant phon_forms and sum the frequencies.
+        # This handles cases where the orth_form is different and has two records.
+        # Paradigms should be read only once, and this code shouldn't be redundant with
+        # the main script. This should be fixed elsewhere.
+        cls.forms['form'] = cls.forms.phon_form
+        dup = cls.forms.duplicated(subset=cls.col_names, keep=False)
+        if dup.any():
+            cls.forms.loc[dup, 'value'] = \
+                cls.forms.loc[dup].groupby(cls.col_names).value.transform(sum)
+            cls.forms.drop_duplicates(subset=cls.col_names, inplace=True)
+
+        cls.forms = cls.forms[['form_id', 'cell', 'lexeme', 'value', 'source']]\
+            .set_index('form_id').sort_index()
+        cls.forms.index.name = "form"
+
+    @classmethod
+    def _read_other_frequencies(cls, name, real=True):
+        """
+        Recover frequency information for cells and lexemes.
+
+        Arguments:
+            name(str): Frequency table to build. Either cells or lexemes.
+            real (bool): Whether to use real frequencies or
+                empty uniform distributions. Defaults to True.
+        """
+        table = px.read_table(name, cls.p, index_col=name[:-1] + "_id")
+
+        # There are up to 3 different situations:
+        # 1. Reading frequencies from the given table.
```

Comment:
Aren't you missing a case where you'd find them in the frequency table (same as with forms ?) ? 
I'm wondering if this and the previous function could be combined (with case (2. building from the forms table) running only if the name isn't forms).

Reply:
> The issue with reading from the frequencies table, is that it is very versatile... So that you never exactly know what you are reading.
The second issue is that the frequency table is in practice unsupported by Paralex (https://gitlab.com/sbeniamine/paralex/-/issues/42), so I could never really see how this would work on real datasets.

Fair enough, let's leave it as that for now

Reply:
> Probably, I was a bit afraid of building a monster function, but I can try !

Have a look, if it makes things simpler then great, otherwise we can keep the current implementation at least for now.